File: subset_sum.cpp
#include <iostream>
#include <vector>

using namespace std;
vector<int> get_array(int length) {
    vector<int> input;
    for (int i = 0; i < length; i++) {
        int inp;
        cin >> inp;
        input.push_back(inp);
    }
    return input;
}

void print_vector(vector<int> vector) {
    cout << "[";
    for (int ele : vector) {
        cout << " " << ele << " ";
    }
    cout << "]" << endl;
}

void find_subset(vector<int> array, int required_sum,
                 vector<int> selected) {
    if (required_sum == 0) {
        print_vector(selected);
        return;
    }
    if (required_sum < 0 or array.empty()) {
        return;
    }
    vector<int> new_vector = {array.begin() + 1, array.end()};
    {
        // include the first element
        vector<int> selected1 = selected;
        selected1.push_back(array[0]);
        find_subset(new_vector, required_sum - array[0], selected1);
    }
    {
        // do not include the first element
        vector<int> selected2 = selected;
        find_subset(new_vector, required_sum, selected2);
    }
}

int main() {
    int length;
    cout << "Length of array: ";
    cin >> length;
    cout << "Array: ";
    vector<int> array = get_array(length);

    int required_sum;
    cout << "Required sum: ";
    cin >> required_sum;

    find_subset(array, required_sum, vector<int>());
}

File: subset_sum.py
def find_subset(array, required_sum, selected=[]):
    """
    Return list of subsets whose sum is equal to required_sum
    """
    # base case
    if required_sum == 0:
        return [selected]
    if required_sum < 0 or not array:
        return False

    result = []

    # include the first element
    result1 = find_subset(array[1:], required_sum - array[0], selected + [array[0]])
    if result1:
        result += result1

    # do not include the first element
    result2 = find_subset(array[1:], required_sum, selected)
    if result2:
        result += result2

    return result


if __name__ == "__main__":
    array = list(map(int, input("Enter array: ").split()))
    required_sum = int(input("Enter required sum: "))

    print(find_subset(array, required_sum))
